docs/proposals/generic-schema-batching-proposal.md

```diff
@@ -0,0 +1,987 @@
+
+**New Endpoint**: `POST /v1/content/:msaId/publish`
+
+**Flow**:
+```
+1. Receive { schemaId, content }
+2. Look up schema via SchemaRegistry
+3. Get adapter via AdapterRegistry (specific or generic)
+4. adapter.validate(content)
+5. transformed = adapter.transform(content, context)
+6. batchConfig = adapter.getBatchConfig()
+7. Queue with schema-specific config
+8. Return 202 Accepted
+```
+
+**Example - DSNP Broadcast** (with adapter):
+```json
+POST /v1/content/12345/publish
+{
+  "schemaId": 17,
+  "content": {
+    "content": {
+      "published": "2025-01-15T10:00:00Z",
+      "content": "Hello DSNP!",
+      "tag": ["#test"]
+    }
+  }
+}
+```
+*Adapter transforms Activity Content → Parquet schema*
+
+**Example - AT Protocol** (without adapter, pass-through):
+```json
+POST /v1/content/12345/publish
+{
+  "schemaId": 100,
+  "content": {
+    "text": "Hello from AT Protocol!",
+    "createdAt": "2025-01-15T10:00:00Z"
+  }
+}
+```
+*Generic adapter passes through pre-formatted content*
+
+#### 6. Adapter Registration Module
+
+**Purpose**: Auto-register all adapters at module initialization.
+
+**Pattern**: NestJS module that injects adapters and registers them with registry.
+
+**Adapters to Create**:
+- `DsnpBroadcastAdapter` (schema 17)
+- `DsnpReplyAdapter` (schema 18)
+- `DsnpReactionAdapter` (schema 4)
+- `DsnpUpdateAdapter` (schema 19)
+- `DsnpProfileAdapter` (schema 6)
+- `DsnpTombstoneAdapter` (schema 16)
+- `GenericSchemaAdapter` (fallback)
+
+**Community Extension**: Third parties can create adapters for their schemas (e.g., `AtProtoPostAdapter`).
+
+### Data Flow Diagram
+
+```
+┌─────────────┐
+│User Request │
+│schemaId:17  │  (DSNP Broadcast)
+└──────┬──────┘
+       │
+       ▼
+┌──────────────────────────┐
+│ Adapter Registry         │
+│ getAdapter(17)           │
+└──────┬───────────────────┘
+       │
+       ▼
+┌──────────────────────────┐
+│ Check Priority-Ordered   │
+│ Adapters for Schema 17   │
+│ • DsnpBroadcastAdapter   │
+│   priority=100 ✓         │
+└──────┬───────────────────┘
+       │
+       ▼
+┌──────────────────────────┐
+│ DsnpBroadcastAdapter     │
+│ • validate(input)        │
+│ • transform(input)       │
+│   └─> createBroadcast()  │
+│ • getBatchConfig()       │
+└──────┬───────────────────┘
+       │
+       ▼
+┌──────────────────────────┐
+│ Transformed Content      │
+│ (Parquet-ready)          │
+└──────┬───────────────────┘
+       │
+       ▼
+┌──────────────────────────┐
+│ Queue for Batching       │
+│ (schema-specific config) │
+└──────────────────────────┘
+```
+
+**vs. AT Protocol (no adapter)**:
+```
+User Request (schemaId:100)
+  → Adapter Registry
+  → No specific adapter found
+  → GenericSchemaAdapter (priority=0)
+  → Pass-through (no transformation)
+  → Queue for Batching
+```
+
+### Benefits of Solution 2
+
+1. **All Benefits of Solution 1** ✅
+2. **Schema-Specific Optimization**: Adapters can implement custom logic, validation, and batch configs per schema
```

Comment:
Devil's advocate:
- custom logic like validation and formatting could be applied in advance of hitting the endpoint, and community adapters used that way would not need support from us.